httpbp/client_middlewares.go

```diff
@@ -0,0 +1,199 @@
+			resp, err = next.RoundTrip(req)
+			err = ClientErrorFromResponse(resp)
+			return resp, err
+		})
+	}
+}
+
+// CircuitBreaker is a middleware that prevents sending requests that are likely
+// to fail based on a configurable failure ratio based on total failures and
+// requests.
+func CircuitBreaker(config ClientConfig) ClientMiddleware {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return &gobreaker.Settings{
+				Interval: 60 * time.Second,
+				ReadyToTrip: func(counts gobreaker.Counts) bool {
+					failureRatio := float64(counts.TotalFailures) / float64(counts.Requests)
+					return counts.Requests >= config.CircuitBreaker.MinRequests && failureRatio >= config.CircuitBreaker.MinFailureRatio
+				},
+			}
+		},
+	}
+	breakers := &sync.Map{}
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (*http.Response, error) {
+			host := req.URL.Hostname()
+
+			// new circuit breakers should rarely get allocated: the pool reduces
+			// garbage collector overhead
+			newBreaker := pool.Get()
+			breaker, loaded := breakers.LoadOrStore(host, newBreaker)
+			if loaded {
+				pool.Put(newBreaker)
+			}
+
+			resp, err := breaker.(*gobreaker.CircuitBreaker).Execute(func() (interface{}, error) {
+				resp, err := next.RoundTrip(req)
+				if err != nil {
+					return nil, err
+				}
+				// circuit break on any HTTP 5xx code
+				if resp.StatusCode >= http.StatusInternalServerError {
+					// read & close to ensure underlying RoundTripper
+					// (http.Transport) is able to re-use the persistent TCP
+					// connection.
+					_, _ = io.ReadAll(resp.Body)
+					err = resp.Body.Close()
+					if err != nil {
+						return nil, err
+					}
+					return nil, err
+				}
+				return resp, nil
+			})
+			if err != nil {
+				return nil, err
+			}
+			return resp.(*http.Response), nil
+		})
+	}
+}
+
+func applyMiddlewares(transport http.RoundTripper, middlewares ...ClientMiddleware) http.RoundTripper {
+	// add middlewares in reverse so the first in the list is the outermost
+	for i := len(middlewares) - 1; i >= 0; i-- {
+		transport = middlewares[i](transport)
+	}
+	return transport
+}
+
+// Retries provides a retry middleware.
+func Retries(retryOptions ...retry.Option) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+
+			if len(retryOptions) == 0 {
+				retryOptions = []retry.Option{retry.Attempts(1)}
+			}
+
+			retrybp.Do(req.Context(), func() error {
+				resp, err = next.RoundTrip(req)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, retryOptions...)
+			if err != nil {
+				return nil, err
+			}
+			return resp, nil
+		})
+	}
+}
+
+// MaxConcurrency is a middleware to ensure that there will be only a maximum
+// number of requests concurrently in-flight at any given time.
+func MaxConcurrency(config ClientConfig) ClientMiddleware {
```

Comment:
As alluded to above, I wouldn't make these top-level functions take in a ClientConfig.  That's passing in far more than is necessary.  These should either be methods on ClientConfig, to make their coupling clear (i.e. so nobody will think they can or should use one without the other) or, even better, make these accept the relevant parameters so they are decoupled from ClientConfig.